pytensor/link/numba/dispatch/basic.py

```diff
@@ -280,104 +275,277 @@ def numba_funcify(op, node=None, storage_map=None, **kwargs):
+    Be sure to read the limitations in the respective docstrings!
+    """
+
+    def decorator(dispatch_func):
+        numba_funcify_and_cache_key.register(op_type)(dispatch_func)
+
+        # Create a wrapper for the legacy dispatcher
+        @wraps(dispatch_func)
+        def dispatch_func_wrapper(*args, **kwargs):
+            func, _key = dispatch_func(*args, **kwargs)
+            # Discard the key for the non-cache version
+            return func
+
+        numba_funcify.register(op_type)(dispatch_func_wrapper)
+
+        return dispatch_func
 
-            return res
+    return decorator
 
+
+def numba_funcify_ensure_cache(op, *args, **kwargs) -> tuple[Callable, str | None]:
+    """Obtain a numba function for an Op and ensure it can be cached by numba.
+
+    If `config.numba__cache` is `True`, and `numba_funcify_and_cache_key` returns a non-None key,
+    the returned function will be wrapped in a python-compiled function that hoists any closures
+    to the global scope. This, together with the NumbaPyTensorCacheLocator ensures numba will use our cache.
+
+    Without this strategy, numba would often consider caches to be invalid. This was always the case for:
+    1. Ops using the custom vectorize intrinsic: Elemwise, Blockwise, RandomVariables
+    2. String generated functions: Alloc, Scan, OpFromGraph, and FunctionGraph itself
+    """
+    if config.numba__cache:
+        jitable_func, cache_key = numba_funcify_and_cache_key(op, *args, **kwargs)
     else:
+        jitable_func, cache_key = numba_funcify(op, *args, **kwargs), None
 
-        @numba_njit
-        def ifelse(cond, *args):
-            if cond:
-                res = args[:n_outs]
-            else:
-                res = args[n_outs:]
+    if cache_key is None:
+        if config.numba__cache and config.compiler_verbose:
+            print(f"{op} of type {type(op)} will not be cached by PyTensor.\n")  # noqa: T201
+        return jitable_func, None
+    else:
+        op_name = op.__class__.__name__
+        cached_func = compile_numba_function_src(
+            src=f"def {op_name}(*args): return jitable_func(*args)",
+            function_name=op_name,
+            global_env=globals() | {"jitable_func": jitable_func},
+            cache_key=cache_key,
+        )
+        return numba_njit(cached_func, cache=True), cache_key
+
+
+def cache_key_for_constant(data):
+    """Create a cache key for a constant value."""
+    if isinstance(data, np.number):
+        return sha256(data).hexdigest()
+    elif isinstance(data, np.ndarray):
+        return hash_from_ndarray(data)
+    elif data is None:
+        return "None"
+    elif isinstance(data, int | float | bool):
+        # These shouldn't all really be np.number, but we keep this branch just in case
+        return str(data)
+    else:
+        # Fallback for arbitrary types
+        return sha256(dumps(data)).hexdigest()
```

Comment:
Maybe a bit nicer to avoid memory use? Probably doesn't matter much in practice though...
```python
hasher = hashlib.sha256()

class HashFile:
    def write(self, b):
        hasher.update(b)

pickle.dump(obj, HashFile())
hasher.hexdigest()
```